### ISSUE TITLE
WebPack 4 breaks Web Worker

### ISSUE THREAD

**Issue by tvogels** [disneyresearch/jeri#12]
Hi,

I'm unable to use the NPM version of JERI as expected. See this [test case](https://github.com/disneyresearch/jeri/files/2670944/jeritest.zip) (run ./run_test.sh) (requires Docker).

```docker
RUN npx create-react-app test
WORKDIR test

# try "npm install --save" as well
ARG INSALLER="yarn add"

RUN ${INSALLER} jeri
RUN ${INSALLER} worker-loader

COPY App.js src/App.js
COPY graphs.png public/
COPY matrix_avg.exr public/

EXPOSE 3000

CMD [ "npm", "run", "start" ]
```

When using `npm install --save`, the script doesn't run, and errors `Cannot read property 'context' of undefined`. When using `yarn add`, the viewer loads and can display PNGs but is stuck on 'Downloading ...' in a tab featuring an EXR image.

**Comment by cbreak-black** [comment]
There seems to be a bug in your EXR Worker, running your example I get the error "ReferenceError: window is not defined exr.worker.js:2:15". Your version shows "/******/ 	var parentHotUpdateCallback = window["webpackHotUpdate"];", while the version on jeri.io doesn't try to access window at all.

**Comment by cbreak-black** [comment]
Maybe... when I let the build run in the node debugger I get pointed to this:

    exception in node_modules/jeri/node_modules/worker-loader/dist/index.js:73
    71
    72   var filename = _loaderUtils2.default.interpolateName(this, options.name || '[hash].worker.js', {
    73     context: options.context || this.options.context,
    74     regExp: options.regExp
    75   });

    > options
    { name: 'exr.worker.js' }
    > this.options
    undefined

**Comment by tvogels** [comment]
@cbreak-black, what should I read from the debugger output?

**Comment by cbreak-black** [comment]
It looks as if the worker-loader tries to get a context member of ```this.options```, which is undefined. I have no idea why it does any of this. But it doesn't look like a bug in JERI, more like something in worker-loader or an other part of the build chain. (Or maybe we don't set options... but I'd expect some form of sane diagnostic from that).

**Comment by tvogels** [comment]
Hm okay. I noticed that jeri.io uses webpack 3 and worker-loader 1, and the current verisons are 4 and 2 respectively. Maybe the rules changed ...

**Comment by cbreak-black** [comment]
Yes, that seems to be the culprit. worker-loader 1.1.1 (instead of 1.1.0) seems to fix the bug we encounter here, but it doesn't seem to be enough to work with webpack 4. Some of your dependencies seem to force the new version onto us.

I'll look into what is required to update our code to use webpack 4 / worker-loader 2, that should circumvent the problem.

**Comment by ja-no** [comment]
FYI: I was able to build this version and install it locally into a different project (react-app) using npm link. It works but EXR images are not being displayed. They are downloaded successfully but then nothing appears in the window. Probably something with exr-parser.worker.js. I'll dig into it.

**Comment by tvogels** [comment]
> FYI: I was able to build this version and install it locally into a different project (react-app) using npm link. It works but EXR images are not being displayed. They are downloaded successfully but then nothing appears in the window. Probably something with exr-parser.worker.js. I'll dig into it.

Hmm are you sure they are downloade successfully?

**Comment by cbreak-black** [comment]
What happens here is that "exr.worker.js" gets downloaded. I have no idea where from. And this file has an error on line 2: ```/******/ 	var parentHotUpdateCallback = window["webpackHotUpdate"];```, the error is ```ReferenceError: window is not defined```.

This file is supposed to be generated out of ```exr-parser.worker.js``` by worker-loader. Chances are worker-loader is doing something wrong.

**Comment by ja-no** [comment]
Yes, I'm getting the `window is not defined` as well.
I'm getting a positive HTTP response so I believe the file is downloaded.

**Comment by cbreak-black** [comment]
The issue is not JERI related I think, but caused by the webpacking done by create-react-app. Our own webpack config isn't considered, it wasn't even bundled in the npm pack.

**Comment by ja-no** [comment]
Hmm... can we `npm eject`? I don't see this option in scripts in package.json. Did you guys remove it manually?

**Comment by cbreak-black** [comment]
After ejecting, you should be able to edit ```config/webpack.config.dev.js```. I tried that, and got it to stop giving that stupid error, but it produced new ones. Why again do we use webpack? :)

**Comment by ja-no** [comment]
How did you eject? There is no script to run `npm run eject`. Or did you do it differently?

**Comment by cbreak-black** [comment]
I did ```npm run eject``` in Thij's example (which ran, so it probably had a script for that). Then I tried removing the ```HotModuleReplacementPlugin`` and adding ```globalObject: `(typeof self !== 'undefined' ? self : this)`,``` as suggested in https://github.com/webpack/webpack/issues/6642, but no success. Maybe the dev server thing is just broken.

**Comment by ja-no** [comment]
After ejecting my test app and adding the globalObject option I'm getting:

`exr-wrap.js:1 Uncaught TypeError: Cannot read property '../../jeri/node_modules/file-loader/dist/cjs.js?name=exr-wrap.wasm!../../jeri/build_npm/exr-wrap/exr-wrap.wasm' of undefined`

I have file-loader 2.0.0 installed. Any idea?

**Comment by cbreak-black** [comment]
So, finally got that weird react app creator thingie to work, for "release" mode. And I hope I didn't forget anything, because this was messy. (I hate webdev... :). I don't know how people can get anything done like this.

- ```npm run eject```
- ```export NODE_ENV=production```
- Change the webpack config file to not break wasm:
  - Add defaultRules next to rules:
    ```
    	defaultRules: [
		{
			type: 'javascript/auto',
			resolve: {}
		},
		{
			test: /\.json$/i,
			type: 'json'
		}
	],
    ```
  - Add '.wasm' to the exclude list for file loader (line 426 in my version of the file):
    ```
    exclude: [/\.(js|mjs|jsx|ts|tsx)$/, /\.html$/, /\.json$/, /\.wasm$/],
    ```
- ```npm install webpack-cli```
- ```./node_modules/.bin/webpack --config config/webpack.config.prod.js```
- ```cp public/graphs.png public/matrix_avg.exr build/```
- ```./node_modules/.bin/serve build/```

**Comment by ja-no** [comment]
Nice!

For me, it was sufficient to eject and make the changes you listed in bullet 3 (default rules and wasm exclude). I didn't need to install webpack-cli and do the rest of what you described.

I'm not getting any errors anymore and the EXR image appears. Thanks for looking into this!

**Comment by ja-no** [comment]
The only annoying thing is that one needs to eject. I wonder if there is a way to change webpack settings without ejecting.

**Comment by tvogels** [comment]
It seems to me like worker-loader etc. are very shaky. It's hard to get it to put the emscripten-related code exactly where it needs to me. Especially because this depends on the user's webpack config. My suggestion would be to let users manually copy the three related files (exr-paser.worker.js, exr-wrap.js, exr-wrap.wasm) to the root of their web server and drop the worker-loader and file-loader stuff.

**Comment by ja-no** [comment]
Would `npm start` find them? E.g. if I put them in the public folder of my react project?

**Comment by tvogels** [comment]
If you put them in the public folder of your react project, they will be served in the web root, so that would work.

**Comment by cbreak-black** [comment]
I updated a ton of dependencies, among them lots of security fixes in 0a9f3a88cf2fb6fa5cf7cfbc03706dc7c1dd6da4, react in 21ed42ab297e077ecccd1940a590965c0958a755 and webpack in 212f0f510de62bf3a543f11a2a257fd12b64ff22.

My stand-alone tests work, so I plan to release this as 2.0 eventually.

**Comment by cbreak-black** [comment]
Webpack 4.31 seems to work somewhat reliably.
